src/String.js

```diff
@@ -12,36 +12,41 @@ class StringT extends Base {
   decode(stream, parent) {
     let length, pos;
 
+    let { encoding } = this;
+    if (typeof encoding === 'function') {
+      encoding = encoding.call(parent, parent) || 'ascii';
+    }
+    let width = encodingWidth(encoding);
+
     if (this.length != null) {
       length = utils.resolveLength(this.length, stream, parent);
     } else {
       let buffer;
       ({buffer, length, pos} = stream);
 
-      while ((pos < length) && (buffer[pos] !== 0x00)) {
-        ++pos;
+      while ((pos < length - width + 1) &&
+        (buffer[pos] !== 0x00 ||
+        (width === 2 && buffer[pos+1] !== 0x00)
```

Comment:
Would it be clearer to change this to

```suggestion
        (buffer[pos+1] !== 0x00 && width === 2)
```

to make it more obvious that the `utf16be` case is already covered by `buffer[pos] !== 0x00` ?

Reply:
We can't do this, because we need to check bounds before we index into the array.